source/1_engine/rex_engine/src/event_system.cpp

```diff
@@ -14,18 +14,75 @@ namespace rex
 {
   namespace event_system
   {
-    rsl::unordered_map<EventType, rsl::vector<EventFunction>>& delegates_map()
+    namespace internal
     {
-      static rsl::unordered_map<EventType, rsl::vector<EventFunction>> delegates_map;
-      return delegates_map;
+      class EventQueue
+      {
+      public:
+        EventQueue()
+            :m_subscriptions()
+            ,m_front()
+            ,m_back()
+            ,m_current(&m_front)
+            ,m_next(&m_back)
+        {
+        }
+
+        void subscribe(EventType type, const EventFunction& function)
+        {
+          m_subscriptions[type].push_back(function);
+        }
+
+        void enqueue(const Event& evt)
+        {
+          m_next->push_back(evt);
+        }
+
+        void fire()
+        {
+          for(auto& evt: *m_current)
+          {
+            const rsl::vector<EventFunction>& delegates = m_subscriptions[evt.type];
+
+            for(const EventFunction& delegate: delegates)
+            {
+              delegate(evt);
+            }
+          }
+
+          m_current->clear();
+        }
+
+        void present()
+        {
+          rsl::vector<Event>* temp = m_current;
```

Comment:
why not use rsl::swap?

Reply:
Dunno, it does the same thing just within a function call.
I'll adjust it if you want.

Reply:
that's why we have functions, to wrap logic so I'd recommend to use it here